### ISSUE TITLE
Neural Networks: Zero To Hero

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9605]
Neural Networks: Zero To Hero<br>
<br>
https://ift.tt/K4xIOC0<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<h1>Neural Networks: Zero to Hero</h1>
<p>A course by Andrej Karpathy on building neural networks, from scratch, in code.</p>
<p>We start with the basics of backpropagation and build up to modern deep neural networks, like GPT. In my opinion language models are an excellent place to learn deep learning, even if your intention is to eventually go to other areas like computer vision because most of what you learn will be immediately transferable. This is why we dive into and focus on languade models.</p>
<p>Prerequisites: solid programming (Python), intro-level math (e.g. derivative, gaussian).</p>
<div>
<p>Learning is easier with others, come say hi in our Discord channel:</p>
<a href="https://discord.gg/3zy8kqD9Cp"><img src="https://dcbadge.vercel.app/api/server/3zy8kqD9Cp"></a>
</div>
<div>
<h2>Syllabus</h2>
<div>
<p>2h25m</p>
<p>This is the most step-by-step spelled-out explanation of backpropagation and training of neural networks. It only assumes basic knowledge of Python and a vague recollection of calculus from high school.</p>
</div>
<div>
<p>1h57m</p>
<p>We implement a bigram character-level language model, which we will further complexify in followup videos into a modern Transformer language model, like GPT. In this video, the focus is on (1) introducing torch.Tensor and its subtleties and use in efficiently evaluating neural networks and (2) the overall framework of language modeling that includes model training, sampling, and the evaluation of a loss (e.g. the negative log likelihood for classification).</p>
</div>
<div>
<p>1h15m</p>
<p>We implement a multilayer perceptron (MLP) character-level language model. In this video we also introduce many basics of machine learning (e.g. model training, learning rate tuning, hyperparameters, evaluation, train/dev/test splits, under/overfitting, etc.).</p>
</div>
<div>
<p>1h55m</p>
<p>We dive into some of the internals of MLPs with multiple layers and scrutinize the statistics of the forward pass activations, backward pass gradients, and some of the pitfalls when they are improperly scaled. We also look at the typical diagnostic tools and visualizations you'd want to use to understand the health of your deep network. We learn why training deep neural nets can be fragile and introduce the first modern innovation that made doing so much easier: Batch Normalization. Residual connections and the Adam optimizer remain notable todos for later video.</p>
</div>
<div>
<p>1h55m</p>
<p>We take the 2-layer MLP (with BatchNorm) from the previous video and backpropagate through it manually without using PyTorch autograd's loss.backward(): through the cross entropy loss, 2nd linear layer, tanh, batchnorm, 1st linear layer, and the embedding table. Along the way, we get a strong intuitive understanding about how gradients flow backwards through the compute graph and on the level of efficient Tensors, not just individual scalars like in micrograd. This helps build competence and intuition around how neural nets are optimized and sets you up to more confidently innovate on and debug modern neural networks.</p>
</div>
<div>
<p>56m</p>
<p>We take the 2-layer MLP from previous video and make it deeper with a tree-like structure, arriving at a convolutional neural network architecture similar to the WaveNet (2016) from DeepMind. In the WaveNet paper, the same hierarchical architecture is implemented more efficiently using causal dilated convolutions (not yet covered). Along the way we get a better sense of torch.nn and what it is and how it works under the hood, and what a typical deep learning development process looks like (a lot of reading of documentation, keeping track of multidimensional tensor shapes, moving between jupyter notebooks and repository code, ...).</p>
</div>
<div>
<p>1h56m</p>
<p>We build a Generatively Pretrained Transformer (GPT), following the paper "Attention is All You Need" and OpenAI's GPT-2 / GPT-3. We talk about connections to ChatGPT, which has taken the world by storm. We watch GitHub Copilot, itself a GPT, help us write a GPT (meta :D!) . I recommend people watch the earlier makemore videos to get comfortable with the autoregressive language modeling framework and basics of tensors and PyTorch nn, which we take for granted in this video.</p>
</div>
<p>ongoing...</p>
</div>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Andrej Karpathy <br>
<br>
July 4, 2023 at 03:21PM
